syncapi/streams/stream_pdu.go

```diff
@@ -222,6 +222,32 @@ func (p *PDUStreamProvider) IncrementalSync(
 		}
 	}
 
+	// If we joined a new room in this sync, make sure we add enough information about it.
+	// This does an "initial sync" for the newly joined rooms
+	newlyJoinedRooms := joinedRooms(req.Response, req.Device.UserID)
+	if len(newlyJoinedRooms) > 0 {
+		// remove already added rooms, as we're doing an "initial sync"
+		for _, x := range newlyJoinedRooms {
+			delete(req.Response.Rooms.Join, x)
+		}
+		r = types.Range{
```

Comment:
You may want this backwards:
```
	r := types.Range{
		From:      to,
		To:        0,
		Backwards: true,
	}
```
... so that you get the most recent timeline events, not the earliest.